File: lib/features/activity_backcountry_ski/models/backcountry_ski_details.dart
import 'package:latlong2/latlong.dart';

import 'package:turbo/features/activities/api.dart' show ActivityGeometry;

/// ATES rating mirroring the server enum.
enum AtesRating { unrated, simple, challenging, complex }

/// 8-point cardinal aspect.
enum Aspect { n, ne, e, se, s, sw, w, nw }

/// Leg of the route.
enum LegKind { ascent, descent, traverse }

class BackcountrySkiDetails {
  final int ascentMeters;
  final int descentMeters;
  final int distanceMeters;
  final int elevationMinMeters;
  final int elevationMaxMeters;
  final AtesRating atesRating;
  final Aspect? dominantAspect;
  final int? varsomRegionId;
  final int? preferredAvalancheMaxLevel;
  final List<AspectShare> aspectMix;
  final List<RouteLeg> legs;

  const BackcountrySkiDetails({
    required this.ascentMeters,
    required this.descentMeters,
    required this.distanceMeters,
    required this.elevationMinMeters,
    required this.elevationMaxMeters,
    required this.atesRating,
    this.dominantAspect,
    this.varsomRegionId,
    this.preferredAvalancheMaxLevel,
    this.aspectMix = const [],
    this.legs = const [],
  });

  Map<String, dynamic> toJson() => {
        'ascentMeters': ascentMeters,
        'descentMeters': descentMeters,
        'distanceMeters': distanceMeters,
        'elevationMinMeters': elevationMinMeters,
        'elevationMaxMeters': elevationMaxMeters,
        'atesRating': atesRating.index,
        'dominantAspect': ?dominantAspect?.index,
        'varsomRegionId': ?varsomRegionId,
        'preferredAvalancheMaxLevel': ?preferredAvalancheMaxLevel,
        'aspectMix': aspectMix.map((a) => a.toJson()).toList(),
        'legs': legs.map((l) => l.toJson()).toList(),
      };

  factory BackcountrySkiDetails.fromJson(Map<String, dynamic> json) =>
      BackcountrySkiDetails(
        ascentMeters: (json['ascentMeters'] as num).toInt(),
        descentMeters: (json['descentMeters'] as num).toInt(),
        distanceMeters: (json['distanceMeters'] as num).toInt(),
        elevationMinMeters: (json['elevationMinMeters'] as num).toInt(),
        elevationMaxMeters: (json['elevationMaxMeters'] as num).toInt(),
        atesRating: AtesRating.values[(json['atesRating'] as num).toInt()],
        dominantAspect: json['dominantAspect'] == null
            ? null
            : Aspect.values[(json['dominantAspect'] as num).toInt()],
        varsomRegionId: (json['varsomRegionId'] as num?)?.toInt(),
        preferredAvalancheMaxLevel:
            (json['preferredAvalancheMaxLevel'] as num?)?.toInt(),
        aspectMix: (json['aspectMix'] as List?)
                ?.cast<Map<String, dynamic>>()
                .map(AspectShare.fromJson)
                .toList() ??
            const [],
        legs: (json['legs'] as List?)
                ?.cast<Map<String, dynamic>>()
                .map(RouteLeg.fromJson)
                .toList() ??
            const [],
      );
}

class AspectShare {
  final Aspect aspect;
  final double fraction;
  const AspectShare({required this.aspect, required this.fraction});

  Map<String, dynamic> toJson() => {
        'aspect': aspect.index,
        'fraction': fraction,
      };

  factory AspectShare.fromJson(Map<String, dynamic> json) => AspectShare(
        aspect: Aspect.values[(json['aspect'] as num).toInt()],
        fraction: (json['fraction'] as num).toDouble(),
      );
}

class RouteLeg {
  final LegKind kind;
  final int startElevationMeters;
  final int endElevationMeters;
  final List<LatLng> polyline;
  const RouteLeg({
    required this.kind,
    required this.startElevationMeters,
    required this.endElevationMeters,
    required this.polyline,
  });

  Map<String, dynamic> toJson() => {
        'kind': kind.index,
        'startElevationMeters': startElevationMeters,
        'endElevationMeters': endElevationMeters,
        'polylineWkt': _toWkt(polyline),
      };

  factory RouteLeg.fromJson(Map<String, dynamic> json) {
    final wkt = json['polylineWkt'] as String;
    final geom = ActivityGeometry.fromServer(wkt: wkt, geometryKind: 'LINESTRING');
    return RouteLeg(
      kind: LegKind.values[(json['kind'] as num).toInt()],
      startElevationMeters: (json['startElevationMeters'] as num).toInt(),
      endElevationMeters: (json['endElevationMeters'] as num).toInt(),
      polyline: geom.coordinates,
    );
  }

  static String _toWkt(List<LatLng> points) {
    if (points.isEmpty) return 'LINESTRING EMPTY';
    final pairs =
        points.map((p) => '${p.longitude} ${p.latitude}').join(', ');
    return 'LINESTRING($pairs)';
  }
}

File: src/Activities/Turbo.Activities.BackcountrySki.Contracts/value/BackcountrySkiDetails.cs
using System.Text.Json.Serialization;

namespace Turboapi.Activities.BackcountrySki.value;

/// <summary>
/// Typed payload of backcountry-ski-specific fields. Every field is
/// typed; nothing JSONB-shaped. Owned-collection-style fields
/// (Aspects, Legs) become their own tables on the read side.
/// </summary>
public sealed record BackcountrySkiDetails
{
    [JsonPropertyName("ascentMeters")] public int AscentMeters { get; init; }
    [JsonPropertyName("descentMeters")] public int DescentMeters { get; init; }
    [JsonPropertyName("distanceMeters")] public int DistanceMeters { get; init; }
    [JsonPropertyName("elevationMinMeters")] public int ElevationMinMeters { get; init; }
    [JsonPropertyName("elevationMaxMeters")] public int ElevationMaxMeters { get; init; }

    [JsonPropertyName("atesRating")] public AtesRating AtesRating { get; init; }
    [JsonPropertyName("dominantAspect")] public Aspect? DominantAspect { get; init; }
    [JsonPropertyName("varsomRegionId")] public int? VarsomRegionId { get; init; }
    [JsonPropertyName("preferredAvalancheMaxLevel")] public short? PreferredAvalancheMaxLevel { get; init; }

    [JsonPropertyName("aspectMix")] public IReadOnlyList<AspectShare> AspectMix { get; init; } = Array.Empty<AspectShare>();
    [JsonPropertyName("legs")] public IReadOnlyList<RouteLeg> Legs { get; init; } = Array.Empty<RouteLeg>();

    [JsonConstructor]
    public BackcountrySkiDetails(
        int ascentMeters,
        int descentMeters,
        int distanceMeters,
        int elevationMinMeters,
        int elevationMaxMeters,
        AtesRating atesRating,
        Aspect? dominantAspect,
        int? varsomRegionId,
        short? preferredAvalancheMaxLevel,
        IReadOnlyList<AspectShare>? aspectMix,
        IReadOnlyList<RouteLeg>? legs)
    {
        AscentMeters = ascentMeters;
        DescentMeters = descentMeters;
        DistanceMeters = distanceMeters;
        ElevationMinMeters = elevationMinMeters;
        ElevationMaxMeters = elevationMaxMeters;
        AtesRating = atesRating;
        DominantAspect = dominantAspect;
        VarsomRegionId = varsomRegionId;
        PreferredAvalancheMaxLevel = preferredAvalancheMaxLevel;
        AspectMix = aspectMix ?? Array.Empty<AspectShare>();
        Legs = legs ?? Array.Empty<RouteLeg>();
    }
}

/// <summary>Fraction of the route that faces a given aspect (0..1).</summary>
public sealed record AspectShare
{
    [JsonPropertyName("aspect")] public Aspect Aspect { get; init; }
    [JsonPropertyName("fraction")] public float Fraction { get; init; }

    [JsonConstructor]
    public AspectShare(Aspect aspect, float fraction)
    {
        Aspect = aspect;
        Fraction = fraction;
    }
}

/// <summary>One segment of the route — typically an ascent, descent, or
/// traverse leg with its own start/end elevation. Geometry of the leg is
/// a substring of the parent route's LineString.</summary>
public sealed record RouteLeg
{
    [JsonPropertyName("kind")] public LegKind Kind { get; init; }
    [JsonPropertyName("startElevationMeters")] public int StartElevationMeters { get; init; }
    [JsonPropertyName("endElevationMeters")] public int EndElevationMeters { get; init; }
    [JsonPropertyName("polylineWkt")] public string PolylineWkt { get; init; }

    [JsonConstructor]
    public RouteLeg(LegKind kind, int startElevationMeters, int endElevationMeters, string polylineWkt)
    {
        Kind = kind;
        StartElevationMeters = startElevationMeters;
        EndElevationMeters = endElevationMeters;
        PolylineWkt = polylineWkt ?? throw new ArgumentNullException(nameof(polylineWkt));
    }
}
